apimachinery: Add a strict YAML and JSON deserializer option

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
pkg/runtime: implement a strict YAML and JSON deserializer

Add a new universal decoder and universal deserializer.
This enables checks for unknown and duplicate fields in input YAML
and JSON data.

Example usage:
```
runtime.DecodeInto(MyCodecFactory.UniversalStrictDecoder(), content, into)
MyCodecFactory.UniversalStrictDeserializer().Decode(content, gvk, into)
```

The same CodecFactory can also return the non-strict variants.

A custom json-iterator API object is used to check for unknown fields.
For duplicate fields the sigs.k8s.io/yaml.YAMLToJSONStrict() function
is used.

Also add:
- Unit tests in json_test.go.
- New error types StrictDecoderError, DuplicateFieldError,
UnknownFieldError.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
xref: https://github.com/kubernetes/community/pull/2977
?

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
apimachinery: Add a strict YAML and JSON deserializer option
```

/assign @liggitt @luxas 
cc @BenTheElder 
/priority important-longterm
/sig api-machinery

neolit123 committed:
pkg/runtime: implement a strict YAML and JSON deserializer

Add a new universal decoder and universal deserializer.
This enables checks for unknown and duplicate fields in input YAML
and JSON data.

Example usage:
runtime.DecodeInto(MyCodecFactory.UniversalStrictDecoder(), content, into)
MyCodecFactory.UniversalStrictDeserializer().Decode(content, gvk, into)

The same CodecFactory can also return the non-strict variants.

A custom json-iterator API object is used to check for unknown fields.
For duplicate fields the sigs.k8s.io/yaml.YAMLToJSONStrict() function
is used.

Also add:
- Unit tests in json_test.go.
- New error types StrictDecoderError, DuplicateFieldError,
UnknownFieldError.

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -159,8 +182,13 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 	if s.yaml {
+		// Check for strict YAML to JSON conversation errors.
+		if s.strict {
+			_, errorYAMLToJSON = yaml.YAMLToJSONStrict(data)
```
Add comment here that error handling is delayed by design, otherwise this seems very odd

luxas replied:
Also comment that this call errors only if (right?) there are duplicate fields, and hence it's put here, before the data variable is mutated with duplicate fields removed from non-strict parsing. That'd clearify things a lot for the reader.

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -159,8 +182,13 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return versioned, actual, nil
 	}
 
+	var errorYAMLToJSON error
```
`YAMLToJSONStrictError`?

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+		// Return an error for unknown JSON fields in strict mode.
+		strictObj := obj.DeepCopyObject()
+		if err := yaml.UnmarshalStrict(data, strictObj); err != nil {
+			return nil, actual, fmt.Errorf("%#v: %v", actual, err)
```
I suggest creating a new error type for this or something so the caller can know whether it was the strict checking or something else that failed.

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -172,6 +183,12 @@ func (f CodecFactory) UniversalDeserializer() runtime.Decoder {
 }
 
+// UniversalStrictDeserializer is the same as UniversalDeserializer expect that it uses a strict Decoder
+// that catches unknown and duplicate fields.
```
that (errors|throws an error) (on|for) unknown and duplicate fields

liggitt replied:
also need to clarify:
* whether decoding continues in the presence of an unknown or duplicate field (is the decoded object usable?)
* what return value signifies an unknown/duplicate field
* how info about the bad fields is extracted from the returned error

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -191,6 +208,18 @@ func (f CodecFactory) UniversalDecoder(versions ...schema.GroupVersion) runtime.
 }
 
+// UniversalStrictDecoder is the same as UniversalDecoder expect that it uses a strict Decoder
+// that catches unknown and duplicate fields.
```
similar nit as above

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+		if errorYAMLToJSON != nil {
+			return nil, actual, fmt.Errorf("%#v: %v", actual, errorYAMLToJSON)
+		}
+		// Return an error for unknown JSON fields in strict mode.
```
nit: "Return an error if there are fields in the provided JSON data that do not exist in the object, or if there are duplicated fields in the JSON file"

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+	// Handle strict mode.
+	if s.strict {
+		// Delayed failure for the strict YAML to JSON conversation,
+		// so that the error output also includes GVK.
```
Add: "This catches the case where duplicate fields are provided in a YAML file"

luxas reviewed (approved):
Thanks @neolit123! Great to see this moving forward :clap:

As a consumer, I want something in between in the meantime that warns the user, not completely exits the application on a failed strict decode, so I think we need to be able to check the error type, and also it'd be nice if we could get some more programmatic information from the error instead of "just" the string, e.g. the gvk metadata for the type decode that failed.

luxas commented:
@neolit123 edited the title/relnote a bit to make it clear that this is optional, and not enforced by default.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -103,12 +109,16 @@ func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
 	for _, d := range serializers {
 		decoders = append(decoders, d.Serializer)
+		if d.StrictSerializer != nil {
+			decodersStrict = append(decodersStrict, d.StrictSerializer)
```
this could end up with an oddly asymmetric decoder if not all serializers provided a strict serializer, which doesn't seem good.

neolit123 replied:
i tough about this, but i don't know how we can ensure that all serializers are strict and the decoders and decodersStrict match.
the main reason for the nil check was that there is another protobuf serializer added to the list that is not setting a StrictSerializer.

on the other hand if one uses:
MyCodecFactory.UniversalStrictDecoder()
this guaranties that only the universalStrict list is going to be used.

liggitt replied:
> the main reason for the nil check was that there is another protobuf serializer added to the list that is not setting a StrictSerializer.

that means that valid protobuf data with no unknown fields could fail strict decoding. that is unexpected.

neolit123 replied:
it seems to me that the protobuf does not have strict mode?
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf_extension.go#L33
V
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go#L154
V
https://github.com/gogo/protobuf/blob/master/proto/decode.go#L335

would it make sense to fallback to a non-strict unmarshall in the case of a nil:
```
if d.StrictSerializer != nil {
	decodersStrict = append(decodersStrict, d.StrictSerializer)
} else {
	decodersStrict = append(decodersStrict, d.Serializer)
}
```

as an extra action, it's also possible to add a StrictSerializer filed here being the same as the regular serializer:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf_extension.go#L39

neolit123 replied:
updated to follow the above strict vs non-strict.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -45,6 +46,16 @@ func NewSerializer(meta MetaFactory, creater runtime.ObjectCreater, typer runtim
 	}
 }
 
+// NewStrictSerializer is the same as NewSerializer expect that it creates a strict JSON serializer.
```
clarify behavior on strict errors, and returned error structure in that case

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+		// Delayed failure for the strict YAML to JSON conversation,
+		// so that the error output also includes GVK.
+		if errorYAMLToJSON != nil {
+			return nil, actual, fmt.Errorf("%#v: %v", actual, errorYAMLToJSON)
```
I expected a structured error we could pull the list of bad fields out of

neolit123 replied:
i think we need to define two error types
- one for duplicate fields
it uses yaml.2 as a backend and just has a slice of regular errors, without fields:
https://github.com/go-yaml/yaml/blob/v2.2.2/yaml.go#L271
https://github.com/go-yaml/yaml/blob/v2/decode.go#L645

- one for unknown fields:
this just uses encoding/json from stdlib, where sadly the error was added as untyped too and only provides text info for a the first failing field it encounters:
https://github.com/golang/go/blob/master/src/encoding/json/decode.go#L736

so the best we can do for now is define our types locally and possibly have them return a list of []errors. but we cannot be extra information in the errors (unless we want to parse Error() string output).

luxas replied:
most importantly, IMO, is to be able to know what problem there was on caller side, so declaring a new error type (e.g. `StrictUnmarshalError`) that then support some `.Cause()` (either "Unknown Fields" or "Duplicate Fields", as a specific string constant type) method would be a good start. If we then additionally can get the fields that's and extra plus but without some fancy string parsing it doesn't seem very easily doable

luxas replied:
...or maybe using two dedicated error types is better

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+			return nil, actual, fmt.Errorf("%#v: %v", actual, errorYAMLToJSON)
+		}
+		// Return an error for unknown JSON fields in strict mode.
+		strictObj := obj.DeepCopyObject()
```
does strict decoding *require* decoding twice and copying the object? that would make it significantly more expensive to use in API server paths

neolit123 replied:
the strict YAMLtoJSON and UnmarsalStrict calls from the k8s.yaml library will always fail on error.

i was thinking about using these calls for the actual result/output too, but this might create discrepancies between json-iter and encoding/json. i think json-iter was choose for a very good reason over encoding/json in api-machinery - possibly case sensitivity and speed?

but if you think that we can just use them as the decoders this will remove the need to make duplicate objects.

liggitt replied:
> i was thinking about using these calls for the actual result/output too, but this might create discrepancies between json-iter and encoding/json. i think json-iter was choose for a very good reason over encoding/json in api-machinery - possibly case sensitivity and speed?

can you give more details about the case-sensitivity behavior? does this library treat `{"metadata":{"Namespace":"..."}}` as an unknown field, or does it fuzzy match it to the "namespace" field? we need it to match the behavior of the thing we're actually using to decode. 

I actually expected to use json-iter's DisallowUnknownFields option for unknown fields, and only make use of the "fail on duplicate fields" option during yaml->json conversion

luxas replied:
right -- alternatively, is there a way we could use json-iter directly, but with some strict configuration of it? does json-iter support erroring on unknown fields? (we already catch the duplicate case with YAMLToJSONStrict in any case)

liggitt replied:
> both ways would be computationally similar as unmarshaling has to be done twice.
> my assumption is that we want to go for 2.

I'd actually want to see the following instead of either of those options:
1. improve the YAMLToJSON convertor to return a list of duplicate key paths encountered while converting. callers can decide whether to fail, warn, or ignore if that list is not empty
2. improve jsoniter decoding to be able to expose a list of unknown field paths encountered while decoding

that would let us avoid any additional copies or unmarshals

luxas replied:
Checked this and it seems to be pretty straightforward actually. We'd create a new `caseSensitiveStrictJsonIterator` and ~duplicate this code using the strict option:
```
// CaseSensitiveJsonIterator returns a jsoniterator API that's configured to be
// case-sensitive when unmarshalling, and otherwise compatible with
// the encoding/json standard library.
func CaseSensitiveJsonIterator() jsoniter.API {
	config := jsoniter.Config{
		EscapeHTML:             true,
		SortMapKeys:            true,
		ValidateJsonRawMessage: true,
		CaseSensitive:          true,
	}.Froze()
	// Force jsoniter to decode number to interface{} via int64/float64, if possible.
	config.RegisterExtension(&customNumberExtension{})
	return config
}

// Private copy of jsoniter to try to shield against possible mutations
// from outside. Still does not protect from package level jsoniter.Register*() functions - someone calling them
// in some other library will mess with every usage of the jsoniter library in the whole program.
// See https://github.com/json-iterator/go/issues/265
var caseSensitiveJsonIterator = CaseSensitiveJsonIterator()
```

Then just use this iterator in `Serializer.Decode` instead of the "normal" iterator. This way we'd avoid two decodes. I'm pretty sure the YAMLToJSONStrict code supports both YAML and JSON inputs (as YAML is a superset of JSON), so we could use this code in ALL cases (not just if you specify YAML) for detecting duplicates in the original data. I think this could be the best option we have so far.

neolit123 replied:
> Checked this and it seems to be pretty straightforward actually.

json iter does suppot it, yes. that was my first attempt.
with the delayed duplicate error approch we can now use https://github.com/json-iterator/go/blob/master/config.go#L135

and create a separate `caseSensitiveStrictJsonIterator` this will avoid desrepencies between the parsers.

neolit123 replied:
@liggitt 
> -improve the YAMLToJSON convertor to accumulate a list of duplicate key paths encountered while converting. callers can decide whether to fail, warn, or ignore if that list is not empty

the yaml to json strict mode comes from the yaml.2 library:
https://github.com/kubernetes-sigs/yaml/blob/master/yaml.go#L124

here:
https://github.com/go-yaml/yaml/blob/v2.2.2/yaml.go#L88

> - improve jsoniter decoding to be able to accumulate a list of unknown field paths encountered while decoding
that would let us avoid any additional copies or unmarshals

unfortunately, these are a couple of upstream changes that may or may not be accepted and i don't want to push for them.

we also have the option to go for 1 but with better error definitions on our side.
i.e. make MyCodecFactory.UniversalStrictDecoder() exit with a meaningful error but not finish the unmarshal.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+		}
+		// Return an error for unknown JSON fields in strict mode.
+		strictObj := obj.DeepCopyObject()
+		if err := yaml.UnmarshalStrict(data, strictObj); err != nil {
```
if we were going to decode twice, I didn't expect to decode into an already-populated object, but a new empty object

neolit123 replied:
i need to figure out how/where to make the copy, if we are going to copy even.

luxas replied:
@neolit123 you might be able to use `scheme.New` for allocating a new, empty object of the same gvk as well

neolit123 replied:
> note that the strictObj could also be a fresh object by passing nil object into UseOrCreateObject.

thanks for the tip.

> @neolit123 you might be able to use scheme.New for allocating a new, empty object of the same gvk as well

hm, do we even have a scheme here passed to the function.

neolit123 commented:
> As a consumer, I want something in between in the meantime that warns the user, not completely exits the application on a failed strict decode, so I think we need to be able to check the error type, and also it'd be nice if we could get some more programmatic information from the error instead of "just" the string, e.g. the gvk metadata for the type decode that failed.

i will definitely include the GVK and properly type the errors.
the lack of proper typing in the backend libraries like encoding/json and yaml.2 is concerning.
they just dump untyped string errors.

currently there are a couple of ways to do this:
1) create two decoders
```
if err := runtime.DecodeInto(MyCodecFactory.UniversalStrictDecoder(), content, into); err != nil {
    return fmt.Println(err) // prints warning
}
if err := runtime.DecodeInto(MyCodecFactory.UniversalDecoder(), content, into); err != nil {
    return err;
}
```
2) check for error types and make the UniversalStrictDecoder not fail:
```
if err := runtime.DecodeInto(MyCodecFactory.UniversalDecoder(), content, into); err != nil {
    switch t := err.(type) {
    case UnknownFieldError:
    case DuplicateFieldError:
        fmt.Println(err)
    default:
        return err
    }
}
```

both ways would be computationally similar as unmarshaling has to be done twice.
my assumption is that we want to go for 2.

luxas commented:
Option 2 is what I prefer, it is way better.

> both ways would be computationally similar as unmarshaling has to be done twice.

See my comment in https://github.com/kubernetes/kubernetes/pull/71589#discussion_r237954162 to avoid doing unmarshal twice.

In any case, I think we need some Go benchmarks to know how much slower the strict decoding is, if we're ever gonna use it in places like the API server where milliseconds matter.

neolit123 commented:
@luxas
> See my comment in #71589 (comment) to avoid doing unmarshal twice.

the problem is that if we enable strict unmarshal for json-iterator it will fail and not trow a warning.
so if we want to go for 2) we still have to unmarshal twice. :\

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,8 +263,32 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+	if s.strict {
+		if yamlToJsonError != nil {
+			str := yamlToJsonError.Error()
+			if strings.Contains(str, `already set in map`) {
```
not a fan of this, but i realized there is no other way to distinguish between duplicate vs unknown field errors.
let me know if you want me to remove the types and always return the generic NewStrictDecoderError instead.

bummer that popular libraries do not return typed errors.

luxas replied:
I'd write a comment about that we eventually want to do this in a better way, but can't atm as the libraries we use don't support structured error returns.

neolit123 commented:
@liggitt @luxas 
updated, i think i addressed the comments.

this ends up being in the lines of:
```
strictDecode() <--- decide to what to do with the error.
decode() <--- can continue to process regularly.
```

please TAL at this part:
https://github.com/kubernetes/kubernetes/pull/71589/files/79837dbe919c2c64d624cbc78ffd2436b3a3b54d#diff-f216f544515d2fd05d66d92c5f95a248
as we may have to remove the custom error types and only leave the StrictDecoderError one.

neolit123 commented:
added unit tests for valid input to the strict decoders.

BenTheElder commented:
scanning for signs of life @luxas @liggitt 🛸

/retest

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/error.go:
```diff
@@ -120,3 +120,77 @@ func IsMissingVersion(err error) bool {
+}
+
+// Gvk returns the GVK that was extacted when Decoding using a strict Decoder.
+func (e *StrictDecoderError) Gvk() schema.GroupVersionKind {
```
nit: maybe consider `GVK()`

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -160,12 +202,35 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
+			data = altered
+		}
+	} else {
+		if s.yaml {
```
I think this could be (maybe better) written as
```go
        var yamlToJSONError error
	if s.strict {
		// In strict mode always pass the data trough the YAMLToJSONStrict converter.
		// This is done to catch duplicate fields for both JSON and YAML. For JSON data,
		// the output would equal the input, unless there is a parsing error such as duplicate fields.
		altered, err := yaml.YAMLToJSONStrict(data)
		if err == nil {
                        // Update data with the sanitized, strict JSON. We update the data variable here, because the original input might have been YAML. For good JSON this is a no-op.
                        data = altered
		} else {
			// Store the error for later use, once a GVK for the input is detected so we can throw a good, structured error.
			yamlToJSONError = err
		}
	}
	// If this is the YAML decoder, and the codec is either non-strict, or have failed to decode in strict mode, convert the provided YAML to JSON in non-strict mode.
	if s.yaml && (!s.strict || yamlToJSONError != nil {
		altered, err := yaml.YAMLToJSON(data)
		if err != nil {
			return nil, nil, err
		}
		data = altered
	}
```

I think this might be easier to read, but IDK. I don't feel strongly

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:
```diff
@@ -213,8 +278,34 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 
-	if err := caseSensitiveJsonIterator.Unmarshal(data, obj); err != nil {
-		return nil, actual, err
+	// Handle strict mode.
```
I'd write a bit more here:
```go
// If the provided decoding mode is strict, first check if there was a duplicate error already earlier in the original JSON 
// or YAML data, and in that case, return it as a DuplicateFieldError. If that's not a problem, decode using the strict 
// json-iter configuration, and return a possible error as UnknownFieldError.
// If non-strict mode is used, just use the "normal" non-strict json-iter decoder.
```

luxas reviewed (commented):
Thanks @neolit123!
This version looks good to me, only minor nits.
Still, if possible, I'd like a benchmark though, but that's optional ;)

luxas commented:
/assign @smarterclayton

neolit123 commented:
i will update the PR with the comments by @luxas on Monday.

neolit123 commented:
> i will update the PR with the comments by @luxas on Monday.

updated.

liggitt commented:
There are three types of behavior we'll eventually want from decoders:
  * ignore duplicate/unknown fields (current behavior)
  * warn on duplicate/unknown fields (useful for surfacing potential issues while keeping API compatibility)
  * error on duplicate/unknown fields (what this PR partially adds)

All of those could be handled uniformly if the decoder returned structured duplicate/unknown field info separately, and the caller decided whether to ignore, warn, or error on it. I'm not sure adding factory APIs to construct alternate decoders with unstructured fail-fast errors for duplicate/unknown fields takes us in the right direction. Would like @smarterclayton's thoughts on the direction of that approach.

neolit123 commented:
> There are three types of behavior we'll eventually want from decoders:

with the current usage of low level libraries, a warning state is not possible without multi-pass unmarshal.
if avoiding multi-pass is a required, said libraries have to be replaced.

> I'm not sure adding factory APIs to construct alternate decoders with unstructured fail-fast errors for duplicate/unknown fields takes us in the right direction

i can still see usage for the separate strict / non strict decoder and the space in between if one wants to handle warnings.

neolit123 commented:
ok, i'm going to leave this PR to the lifecycle bots.
if there is a need for the PR to merge before it rots we can rebase and update.

neolit123 commented:
closing in favor of: #72883
